### PR TITLE
Implement HTTP middleware to official recommendation #94

### PR DESCRIPTION
**Pull request by benmccallum**
PR for issue logged at: https://github.com/graphql-dotnet/server/issues/94

### CODE REVIEW

File: src/AspNetCore/GraphQLHttpMiddleware.cs [hunk 1]
```diff
@@ -54,7 +56,27 @@ private bool IsGraphQlRequest(HttpContext context)
 
         private async Task ExecuteAsync(HttpContext context, ISchema schema)
         {
-            var request = Deserialize<GraphQLQuery>(context.Request.Body);
+            // Handle requests as per recommendation at http://graphql.org/learn/serving-over-http/
+            var request = new GraphQLQuery();
+            if (context.Request.Method.Equals("GET", StringComparison.OrdinalIgnoreCase))
```
Comment (pekkah) [inline src/AspNetCore/GraphQLHttpMiddleware.cs]: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpmethods?view=aspnetcore-2.0

File: src/AspNetCore/GraphQLHttpMiddleware.cs [hunk 2]
```diff
@@ -54,7 +56,27 @@ private bool IsGraphQlRequest(HttpContext context)
...
+            {
+                var qs = context.Request.Query;
+                request.Query = qs.TryGetValue("query", out StringValues queryValues) ? queryValues[0] : null;
+                request.Variables = qs.TryGetValue("variables", out StringValues variablesValues) ? JObject.Parse(variablesValues[0]) : null;
+                request.OperationName = qs.TryGetValue("operationName", out StringValues operationNameValues) ? operationNameValues[0] : null;
+            }
+            else if (context.Request.Method.Equals("POST", StringComparison.OrdinalIgnoreCase))
```
Comment (pekkah) [inline src/AspNetCore/GraphQLHttpMiddleware.cs]: See above

File: src/AspNetCore/GraphQLHttpMiddleware.cs [hunk 3]
```diff
@@ -54,7 +56,27 @@ private bool IsGraphQlRequest(HttpContext context)
...
+            {
+                switch (context.Request.ContentType)
```
Comment (pekkah) [inline src/AspNetCore/GraphQLHttpMiddleware.cs]: Does this handle encodings as well?

Comment (benmccallum) [reply src/AspNetCore/GraphQLHttpMiddleware.cs]: Not exactly sure what you mean by encodings. Could you give an example?

**Comment by benmccallum** [comment]
Hey @pekkah , keen to get this through. See my question above.

**Comment by pekkah** [comment]
If the content type includes the charset=UTF-8 will this still work? I haven't checked if the ASP.NET Core parses the contenttype with charset or not..

**Comment by benmccallum** [comment]
Oh I see what you mean, I'll have a look now.

**Comment by benmccallum** [comment]
You're right. The charset and whatever else is passed on the end of the ContentType header does come through and will break my comparison. Good pickup. Code reviews FTW.

Did a bit of [digging](https://github.com/aspnet/Mvc/blob/760c8f38678118734399c58c2dac981ea6e47046/test/WebSites/BasicWebSite/Formatters/CustomFormatter.cs) and found `System.Net.Http.Headers.MediaTypeHeaderValue` which that can normalize it.

Also added a missing scenario, handling the POST request with ?query={...} supplied. In this case it should be handled as the GET request is.

Few concerns I have now are:

1. POST with broken `ContentType: application/json; charset=myNonExistantCharset;` will skip the parsing, go to the DocumentExecutor with empty GraphQLRequest and come back with something like "query must be supplied". Should we short-circuit and return a BadRequest with relevant message?
2. POST with empty ContentType or anything we don't handle, same thing? Short-circuit with a BadRequest?

**Comment by pekkah** [comment]
For 1. and 2. Bad request sounds ok.

**Comment by benmccallum** [comment]
@pekkah, done mate. Over to you.

**Comment by pekkah** [comment]
Some of those nested if,switches could use some cleanup but that can be a chore before stable release.

**Comment by pekkah** [comment]
Just noticed that you created this against the master branch. Could you change it to develop as that's where the next version is brewing.

**Comment by benmccallum** [comment]
Sure, no problem.

**Pull request closed without merging**